sw/otbn/code-snippets/barrett384.S

```diff
@@ -0,0 +1,254 @@
+
+  // Compute r1 = x % 2^385
+  // r1 = [w22, w21] = x % 2^385 = x & (2^3851) = ([w9, w8] * [w13,w12]) & [w30, w31]
+  // => max length r1: 385 bits
+  BN.MOV w21, w16
+  BN.AND w22, w17, w30
+
+  // Compute q1 = x >> 383
+  // q1 = [w9, w8] = [w18, w17, w16] >> 383  = [w18, w17] >> 127
+  // => max length q1: 385 bits
+  BN.RSHI w9, w18, w31 >> 127
+  BN.RSHI w8, w17, w18 >> 127
+
+  // Compute q2 = q1*u
+  // Instead of full q2 (which would be up to 770 bits) we ignore the MSb of u and the MSb of q1 and correct this later.
+  // This allows using a smaller 384x384 multiplier. For special modili where the lower half of the second limb of u is
+  // zero (e.g. p384) this can be further optimized by only considering limb 0 of u and use a 384x256 multiplication.
+  // max. length q2': 768 bit
+  // q2' = q1[383:0]*u[383:0] = [w18, w17, w16] = [w9, w8] * [w15, w14]
+  BN.MOV w10, w14
+  BN.MOV w11, w15
+  JAL x1, mul384
+
+  // q3 = q2 >> 385
+  // In this step, the compensation for the neglected MSbs of q1 and u is carried out underway. To add them in
+  // the q2 domain, they would have to be left shifted by 384 bit first. To directly add them we first shift q2' by
+  // 384 bit to the right, perform the additions, and shift the result another bit to the right. The additions cannot
+  // overflow due to leading zeros after shift.
+  // q2'' = q2' >> 384 = [w20, w19] = [w18, w17, w16] >> 384 = [w18, w17] >> 128
+  BN.RSHI w20, w18, w31 >> 128
+  BN.RSHI w19, w17, w18 >> 128
+  // Add q1. This is unconditional since MSb of u is always 1. This cannot overflow due to leading zeros.
+  // q2''' = q2'' + q1 = [w20, w19] = [w20, w19] + [w8, w9]
+  BN.ADD w19, w19, w8
+  BN.ADDC w20, w20, w9
+  // Conditionally add u (without leading 1) in case of MSb of x being set.
+  // This is the "real" q2 but shifted by 384 bits to the right. This cannot overflow due to leading zeros */
+  // q2'''' = x[767]?q2'''+u[383:0]:q2''' = [w20, w19] + [w25, w24] = q2 >> 384
+  BN.ADD w19, w19, w24
+  BN.ADDC w20, w20, w25
+  // finally this gives q3 by shifting the remain bit to the right
+  // q3 = q2 >> 385 = q2'''' >> 1 = [w9, w8] = [w20, w19] >> 1
+  BN.RSHI w9, w20, w31 >> 1
+  BN.RSHI w8, w19, w20 >> 1
+
+  // Compute r2 = q3 * m % 2^385.
+  // => max. length r2: 385 bit
+  // q3*m = [w18, w17, w16] = [w9, w8] * [w13,w12]
+  BN.MOV w10, w12
+  BN.MOV w11, w13
+  JAL x1, mul384
+  // r2 = [w17, w16] = q3*m%2^385 = [w18, w17, w16] & (2^385-1) = [w17&(2^129-1), w16] = [w17&w30, w16]
+  BN.AND w17, w17, w30
+
+  // Compute r = r1 - r2.
+  // => max. length r: 385 bit
+  // [w22,w21] = [w22,w21] - [w17,w16]
+  BN.SUB w21, w21, w16
+  BN.SUBB w22, w22, w17
+
+  // Reuse old 129-bit mask to get single 1 at bit pos 129
+  // 2^129 = 2^129-1 +1 = w[30] <= [w30] +1
+  BN.ADDI w30, w30, 1
+
+  // Conditionally add 2^385 to r if r is negative.
+  // r = [w22, w21] <= (r < 0)?r + 2^385 = [w22, w21] + 2^385 = [w22 + 2^129, w21]
+  BN.ADD w30, w30, w22
+  BN.SEL w22, w30, w22, c ;
```

Comment:
Again, the flag group needs capitalising here and on the next few lines. Also, we don't support `;` as a comment character.

Reply:
done